### PR TITLE
[Keyboard] Add dnworks 997pt3

### PR DESCRIPTION
**Pull request by itsme-zeix**
## Description

Added support for dnworks 997.3 TKL and its variants. Thank you!

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/dnworks/9973/keymaps/default/keymap.c [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  KC_LCTL, KC_LGUI, KC_LALT,                            KC_SPC,                             KC_RALT, KC_RGUI, KC_RCTL,                   KC_LEFT, KC_DOWN, KC_RIGHT),
+  
+[1] = LAYOUT(
+  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,                   KC_TRNS, KC_TRNS, KC_TRNS,
+  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,          KC_TRNS, KC_TRNS, KC_TRNS,
+  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,          KC_TRNS, KC_TRNS, KC_TRNS,
+  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,          KC_TRNS,         
+  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,                            KC_TRNS,
+  KC_TRNS, KC_TRNS, KC_TRNS,                            KC_TRNS,                            KC_TRNS, KC_TRNS, KC_TRNS,                   KC_TRNS, KC_TRNS, KC_TRNS),
```
[the remark this answers is not in the document]
Comment (itsme-zeix) [reply keyboards/dnworks/9973/keymaps/default/keymap.c:34]: TIL! Thank you for the review!

**Review by fauxpark — APPROVED** [review]

File: keyboards/dnworks/9973/info.json [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+                { "matrix": [5, 2], "w": 1.5, "x": 2.5, "y": 5.5 },
+                { "matrix": [5, 6], "w": 7, "x": 4, "y": 5.5 },
+                { "matrix": [5, 10], "w": 1.5, "x": 11, "y": 5.5 },
+                { "matrix": [5, 11], "x": 12.5, "y": 5.5 },
+                { "matrix": [5, 12], "w": 1.5, "x": 13.5, "y": 5.5 },
+                { "matrix": [5, 14], "x": 15.25, "y": 5.5 },
+                { "matrix": [5, 15], "x": 16.25, "y": 5.5 },
+                { "matrix": [5, 16], "x": 17.25, "y": 5.5 }
+            ]
+        }
```
Comment (dunk2k) [inline keyboards/dnworks/9973/info.json:119]:
```suggestion
        },
        "LAYOUT_tkl_ansi_tsangan": {
            "layout": [
                { "matrix": [0, 0], "x": 0, "y": 0 },
                { "matrix": [0, 2], "x": 2, "y": 0 },
                { "matrix": [0, 3], "x": 3, "y": 0 },
                { "matrix": [0, 4], "x": 4, "y": 0 },
                { "matrix": [0, 5], "x": 5, "y": 0 },
                { "matrix": [0, 6], "x": 6.5, "y": 0 },
                { "matrix": [0, 7], "x": 7.5, "y": 0 },
                { "matrix": [0, 8], "x": 8.5, "y": 0 },
                { "matrix": [0, 9], "x": 9.5, "y": 0 },
                { "matrix": [0, 10], "x": 11, "y": 0 },
                { "matrix": [0, 11], "x": 12, "y": 0 },
                { "matrix": [0, 12], "x": 13, "y": 0 },
                { "matrix": [0, 13], "x": 14, "y": 0 },
                { "matrix": [0, 14], "x": 15.25, "y": 0 },
                { "matrix": [0, 15], "x": 16.25, "y": 0 },
                { "matrix": [0, 16], "x": 17.25, "y": 0 },
                { "matrix": [1, 0], "x": 0, "y": 1.5 },
                { "matrix": [1, 1], "x": 1, "y": 1.5 },
                { "matrix": [1, 2], "x": 2, "y": 1.5 },
                { "matrix": [1, 3], "x": 3, "y": 1.5 },
                { "matrix": [1, 4], "x": 4, "y": 1.5 },
                { "matrix": [1, 5], "x": 5, "y": 1.5 },
                { "matrix": [1, 6], "x": 6, "y": 1.5 },
                { "matrix": [1, 7], "x": 7, "y": 1.5 },
                { "matrix": [1, 8], "x": 8, "y": 1.5 },
                { "matrix": [1, 9], "x": 9, "y": 1.5 },
                { "matrix": [1, 10], "x": 10, "y": 1.5 },
                { "matrix": [1, 11], "x": 11, "y": 1.5 },
                { "matrix": [1, 12], "x": 12, "y": 1.5 },
                { "matrix": [1, 13], "w": 2, "x": 13, "y": 1.5 },
                { "matrix": [1, 14], "x": 15.25, "y": 1.5 },
                { "matrix": [1, 15], "x": 16.25, "y": 1.5 },
                { "matrix": [1, 16], "x": 17.25, "y": 1.5 },
                { "matrix": [2, 0], "w": 1.5, "x": 0, "y": 2.5 },
                { "matrix": [2, 1], "x": 1.5, "y": 2.5 },
                { "matrix": [2, 2], "x": 2.5, "y": 2.5 },
                { "matrix": [2, 3], "x": 3.5, "y": 2.5 },
                { "matrix": [2, 4], "x": 4.5, "y": 2.5 },
                { "matrix": [2, 5], "x": 5.5, "y": 2.5 },
                { "matrix": [2, 6], "x": 6.5, "y": 2.5 },
                { "matrix": [2, 7], "x": 7.5, "y": 2.5 },
                { "matrix": [2, 8], "x": 8.5, "y": 2.5 },
                { "matrix": [2, 9], "x": 9.5, "y": 2.5 },
                { "matrix": [2, 10], "x": 10.5, "y": 2.5 },
                { "matrix": [2, 11], "x": 11.5, "y": 2.5 },
                { "matrix": [2, 12], "x": 12.5, "y": 2.5 },
                { "matrix": [2, 13], "w": 1.5, "x": 13.5, "y": 2.5 },
                { "matrix": [2, 14], "x": 15.25, "y": 2.5 },
                { "matrix": [2, 15], "x": 16.25, "y": 2.5 },
                { "matrix": [2, 16], "x": 17.25, "y": 2.5 },
                { "matrix": [3, 0], "w": 1.75, "x": 0, "y": 3.5 },
                { "matrix": [3, 1], "x": 1.75, "y": 3.5 },
                { "matrix": [3, 2], "x": 2.75, "y": 3.5 },
                { "matrix": [3, 3], "x": 3.75, "y": 3.5 },
                { "matrix": [3, 4], "x": 4.75, "y": 3.5 },
                { "matrix": [3, 5], "x": 5.75, "y": 3.5 },
                { "matrix": [3, 6], "x": 6.75, "y": 3.5 },
                { "matrix": [3, 7], "x": 7.75, "y": 3.5 },
                { "matrix": [3, 8], "x": 8.75, "y": 3.5 },
                { "matrix": [3, 9], "x": 9.75, "y": 3.5 },
                { "matrix": [3, 10], "x": 10.75, "y": 3.5 },
                { "matrix": [3, 11], "x": 11.75, "y": 3.5 },
                { "matrix": [3, 12], "w": 2.25, "x": 12.75, "y": 3.5 },
                { "matrix": [4, 0], "w": 2.25, "x": 0, "y": 4.5 },
                { "matrix": [4, 2], "x": 2.25, "y": 4.5 },
                { "matrix": [4, 3], "x": 3.25, "y": 4.5 },
                { "matrix": [4, 4], "x": 4.25, "y": 4.5 },
                { "matrix": [4, 5], "x": 5.25, "y": 4.5 },
                { "matrix": [4, 6], "x": 6.25, "y": 4.5 },
                { "matrix": [4, 7], "x": 7.25, "y": 4.5 },
                { "matrix": [4, 8], "x": 8.25, "y": 4.5 },
                { "matrix": [4, 9], "x": 9.25, "y": 4.5 },
                { "matrix": [4, 10], "x": 10.25, "y": 4.5 },
                { "matrix": [4, 11], "x": 11.25, "y": 4.5 },
                { "matrix": [4, 12], "w": 2.75, "x": 12.25, "y": 4.5 },
                { "matrix": [4, 15], "x": 16.25, "y": 4.5 },
                { "matrix": [5, 0], "w": 1.5, "x": 0, "y": 5.5 },
                { "matrix": [5, 1], "x": 1.5, "y": 5.5 },
                { "matrix": [5, 2], "w": 1.5, "x": 2.5, "y": 5.5 },
                { "matrix": [5, 6], "w": 7, "x": 4, "y": 5.5 },
                { "matrix": [5, 10], "w": 1.5, "x": 11, "y": 5.5 },
                { "matrix": [5, 11], "x": 12.5, "y": 5.5 },
                { "matrix": [5, 12], "w": 1.5, "x": 13.5, "y": 5.5 },
                { "matrix": [5, 14], "x": 15.25, "y": 5.5 },
                { "matrix": [5, 15], "x": 16.25, "y": 5.5 },
                { "matrix": [5, 16], "x": 17.25, "y": 5.5 }
            ]
        },
        "LAYOUT_tkl_ansi_wkl": {
            "layout": [
                { "matrix": [0, 0], "x": 0, "y": 0 },
                { "matrix": [0, 2], "x": 2, "y": 0 },
                { "matrix": [0, 3], "x": 3, "y": 0 },
                { "matrix": [0, 4], "x": 4, "y": 0 },
                { "matrix": [0, 5], "x": 5, "y": 0 },
                { "matrix": [0, 6], "x": 6.5, "y": 0 },
                { "matrix": [0, 7], "x": 7.5, "y": 0 },
                { "matrix": [0, 8], "x": 8.5, "y": 0 },
                { "matrix": [0, 9], "x": 9.5, "y": 0 },
                { "matrix": [0, 10], "x": 11, "y": 0 },
                { "matrix": [0, 11], "x": 12, "y": 0 },
                { "matrix": [0, 12], "x": 13, "y": 0 },
                { "matrix": [0, 13], "x": 14, "y": 0 },
                { "matrix": [0, 14], "x": 15.25, "y": 0 },
                { "matrix": [0, 15], "x": 16.25, "y": 0 },
                { "matrix": [0, 16], "x": 17.25, "y": 0 },
                { "matrix": [1, 0], "x": 0, "y": 1.5 },
                { "matrix": [1, 1], "x": 1, "y": 1.5 },
                { "matrix": [1, 2], "x": 2, "y": 1.5 },
                { "matrix": [1, 3], "x": 3, "y": 1.5 },
                { "matrix": [1, 4], "x": 4, "y": 1.5 },
                { "matrix": [1, 5], "x": 5, "y": 1.5 },
                { "matrix": [1, 6], "x": 6, "y": 1.5 },
                { "matrix": [1, 7], "x": 7, "y": 1.5 },
                { "matrix": [1, 8], "x": 8, "y": 1.5 },
                { "matrix": [1, 9], "x": 9, "y": 1.5 },
                { "matrix": [1, 10], "x": 10, "y": 1.5 },
                { "matrix": [1, 11], "x": 11, "y": 1.5 },
                { "matrix": [1, 12], "x": 12, "y": 1.5 },
                { "matrix": [1, 13], "w": 2, "x": 13, "y": 1.5 },
                { "matrix": [1, 14], "x": 15.25, "y": 1.5 },
                { "matrix": [1, 15], "x": 16.25, "y": 1.5 },
                { "matrix": [1, 16], "x": 17.25, "y": 1.5 },
                { "matrix": [2, 0], "w": 1.5, "x": 0, "y": 2.5 },
                { "matrix": [2, 1], "x": 1.5, "y": 2.5 },
                { "matrix": [2, 2], "x": 2.5, "y": 2.5 },
                { "matrix": [2, 3], "x": 3.5, "y": 2.5 },
                { "matrix": [2, 4], "x": 4.5, "y": 2.5 },
                { "matrix": [2, 5], "x": 5.5, "y": 2.5 },
                { "matrix": [2, 6], "x": 6.5, "y": 2.5 },
                { "matrix": [2, 7], "x": 7.5, "y": 2.5 },
                { "matrix": [2, 8], "x": 8.5, "y": 2.5 },
                { "matrix": [2, 9], "x": 9.5, "y": 2.5 },
                { "matrix": [2, 10], "x": 10.5, "y": 2.5 },
                { "matrix": [2, 11], "x": 11.5, "y": 2.5 },
                { "matrix": [2, 12], "x": 12.5, "y": 2.5 },
                { "matrix": [2, 13], "w": 1.5, "x": 13.5, "y": 2.5 },
                { "matrix": [2, 14], "x": 15.25, "y": 2.5 },
                { "matrix": [2, 15], "x": 16.25, "y": 2.5 },
                { "matrix": [2, 16], "x": 17.25, "y": 2.5 },
                { "matrix": [3, 0], "w": 1.75, "x": 0, "y": 3.5 },
                { "matrix": [3, 1], "x": 1.75, "y": 3.5 },
                { "matrix": [3, 2], "x": 2.75, "y": 3.5 },
                { "matrix": [3, 3], "x": 3.75, "y": 3.5 },
                { "matrix": [3, 4], "x": 4.75, "y": 3.5 },
                { "matrix": [3, 5], "x": 5.75, "y": 3.5 },
                { "matrix": [3, 6], "x": 6.75, "y": 3.5 },
                { "matrix": [3, 7], "x": 7.75, "y": 3.5 },
                { "matrix": [3, 8], "x": 8.75, "y": 3.5 },
                { "matrix": [3, 9], "x": 9.75, "y": 3.5 },
                { "matrix": [3, 10], "x": 10.75, "y": 3.5 },
                { "matrix": [3, 11], "x": 11.75, "y": 3.5 },
                { "matrix": [3, 12], "w": 2.25, "x": 12.75, "y": 3.5 },
                { "matrix": [4, 0], "w": 2.25, "x": 0, "y": 4.5 },
                { "matrix": [4, 2], "x": 2.25, "y": 4.5 },
                { "matrix": [4, 3], "x": 3.25, "y": 4.5 },
                { "matrix": [4, 4], "x": 4.25, "y": 4.5 },
                { "matrix": [4, 5], "x": 5.25, "y": 4.5 },
                { "matrix": [4, 6], "x": 6.25, "y": 4.5 },
                { "matrix": [4, 7], "x": 7.25, "y": 4.5 },
                { "matrix": [4, 8], "x": 8.25, "y": 4.5 },
                { "matrix": [4, 9], "x": 9.25, "y": 4.5 },
                { "matrix": [4, 10], "x": 10.25, "y": 4.5 },
                { "matrix": [4, 11], "x": 11.25, "y": 4.5 },
                { "matrix": [4, 12], "w": 2.75, "x": 12.25, "y": 4.5 },
                { "matrix": [4, 15], "x": 16.25, "y": 4.5 },
                { "matrix": [5, 0], "w": 1.5, "x": 0, "y": 5.5 },
                { "matrix": [5, 2], "w": 1.5, "x": 2.5, "y": 5.5 },
                { "matrix": [5, 6], "w": 7, "x": 4, "y": 5.5 },
                { "matrix": [5, 10], "w": 1.5, "x": 11, "y": 5.5 },
                { "matrix": [5, 12], "w": 1.5, "x": 13.5, "y": 5.5 },
                { "matrix": [5, 14], "x": 15.25, "y": 5.5 },
                { "matrix": [5, 15], "x": 16.25, "y": 5.5 },
                { "matrix": [5, 16], "x": 17.25, "y": 5.5 }
```

File: keyboards/dnworks/9973/info.json [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    "indicators": {
+        "caps_lock": "GP27",
+        "scroll_lock": "GP1",
+        "on_state": 1
+    },
+    "diode_direction": "COL2ROW",
+    "matrix_pins": {
+        "rows": ["GP2", "GP3", "GP4", "GP29", "GP12", "GP13"],
+        "cols": ["GP28", "GP26", "GP25", "GP24", "GP23", "GP22", "GP21", "GP20", "GP19", "GP18", "GP17", "GP16", "GP15", "GP14", "GP6", "GP5", "GP0"]
+    },
```
Comment (dunk2k) [inline keyboards/dnworks/9973/info.json:27]:
```suggestion
    },
    "community_layouts": ["tkl_ansi_tsangan"],
```

File: keyboards/dnworks/9973/info.json [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    "layouts": {
```
Comment (dunk2k) [inline keyboards/dnworks/9973/info.json:28]: @itsme-zeix 
Regarding physical layout of this board, does production PCB (can only find images of prototype) have ISO layout support? i.e. could equivalent ISO layout be added

**Review by dunk2k — CHANGES_REQUESTED** [review]

File: keyboards/dnworks/9973/info.json (hunk 4 above)
Comment (itsme-zeix) [reply keyboards/dnworks/9973/info.json:28]: Yes, it does support ISO! The ISO Enter is mapped to the ANSI Enter [3, 12]. ISO Non-US-Hash is mapped to the ANSI Pipe key [2, 13]. In that case, would a community layout need to be added for ISO, presumably adjusted for key sizing/positions?

On that note, how do you guys usually generate the matrix diagram and community layouts? Would that be by hand or is there some tool available?

File: keyboards/dnworks/9973/rules.mk [hunk 5]
```diff
@@ -0,0 +1 @@
+# This file intentionally left blank
```
Comment (dunk2k) [inline keyboards/dnworks/9973/rules.mk:1]: add EoF
```suggestion
# This file intentionally left blank

```

File: keyboards/dnworks/9973/info.json [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        "LAYOUT": {
```
Comment (dunk2k) [inline keyboards/dnworks/9973/info.json:29]:
```suggestion
        "LAYOUT_all": {
```

File: keyboards/dnworks/9973/keymaps/default/keymap.c [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+[0] = LAYOUT(
```
Comment (dunk2k) [inline keyboards/dnworks/9973/keymaps/default/keymap.c:20]:
```suggestion
[0] = LAYOUT_all(
```

File: keyboards/dnworks/9973/keymaps/via/keymap.c [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+[0] = LAYOUT(
```
Comment (dunk2k) [inline keyboards/dnworks/9973/keymaps/via/keymap.c:20]:
```suggestion
[0] = LAYOUT_all(
```

**Review by dunk2k — CHANGES_REQUESTED** [review]

File: keyboards/dnworks/9973/info.json (hunk 4 above)
Comment (dunk2k) [reply keyboards/dnworks/9973/info.json:28]: > Yes, it does support ISO! The ISO Enter is mapped to the ANSI Enter [3, 12]. ISO Non-US-Hash is mapped to the ANSI Pipe key [2, 13]. In that case, would a community layout need to be added for ISO, presumably adjusted for key sizing/positions?

edited related suggestion for ease 😉

> On that note, how do you guys usually generate the matrix diagram and community layouts?

Community Layouts and their templated information are available in directory `layouts/default/`:
- Key sequence/order, key spacings, and key widths for a given Community Layout are found in respective layout's `info.json` file
  - visual representation of a Community Layout can be copy-pasted from `layout.json` into "Raw Data" tab of [KLE](http://www.keyboard-layout-editor.com/#/) 
- ASCII diagram templates, for `matrix_diagram.md`, found in `layouts/default/readme.md`

> Would that be by hand or is there some tool available?

To my knowledge, and happy to be corrected, is done by hand - a GUI tool and/or QMK CLI command that generates multiple layout information to a .json file would be 🤩

Comment (itsme-zeix) [reply keyboards/dnworks/9973/info.json:28]: No worries, thank you for the changes made!

Noted on the community layouts/matrix diagrams - I will try to incorporate these in future PRs. Also, never knew about the configurator's preview function, this sounds really, really useful. Perhaps I may come up with some tool as a side project to aid in the firmware generation after my finals (assuming enough motivation). Thanks for the review again!

File: keyboards/dnworks/9973/info.json [hunk 9]
```diff
@@ -0,0 +1,478 @@
...
+                { "matrix": [4, 15], "x": 16.25, "y": 4.5 },
+                { "matrix": [5, 0], "w": 1.5, "x": 0, "y": 5.5 },
+                { "matrix": [5, 2], "w": 1.5, "x": 2.5, "y": 5.5 },
+                { "matrix": [5, 6], "w": 7, "x": 4, "y": 5.5 },
+                { "matrix": [5, 10], "w": 1.5, "x": 11, "y": 5.5 },
+                { "matrix": [5, 12], "w": 1.5, "x": 13.5, "y": 5.5 },
+                { "matrix": [5, 14], "x": 15.25, "y": 5.5 },
+                { "matrix": [5, 15], "x": 16.25, "y": 5.5 },
+                { "matrix": [5, 16], "x": 17.25, "y": 5.5 }
+    }
```
Comment (dunk2k) [inline keyboards/dnworks/9973/info.json:479]: Linting/syntax correction
```suggestion
            ]
        }
    }
```

**Review by dunk2k — CHANGES_REQUESTED** [review]

**Review by waffle87 — APPROVED** [review]

**Pull request merged**